File: src/main/scala/assignment3/e0/ParticleBuffer.scala
package assignment3.e0

import scala.collection.mutable
import scala.collection.mutable.ListBuffer

class ParticleBuffer() {

    private var particles: mutable.ListBuffer[Particle] = new mutable.ListBuffer()

     def getNewParticleAvail: Particle = synchronized {
       if (particles.nonEmpty)
         particles.remove(1)
       else null
     }

     def notifyNewParticle(p: Particle): ListBuffer[Particle] = synchronized{
       particles += p
     }
}

File: src/main/scala/assignment3/e0/ParticleBuffer.java
package assignment3.e0;

import java.util.LinkedList;
import java.util.Optional;

public class ParticleBuffer {

	private LinkedList<Particle> particles;
	
	public ParticleBuffer() {
		particles = new LinkedList<Particle>();
	}
	
	synchronized Optional<Particle> getNewParticleAvail() {
		if (particles.size() > 0) {
			return Optional.of(particles.remove());
		} else {
			return Optional.empty();
		}
	}
	
	synchronized void notifyNewParticle(Particle p) {
		particles.add(p);
	}
}
